source/extensions/http/custom_response/redirect_policy/redirect_policy.cc

```diff
@@ -0,0 +1,177 @@
+    auto const cr_code = ::Envoy::Http::Utility::getResponseStatusOrNullopt(headers);
+    if (!cr_code.has_value() || (*cr_code < 100 || *cr_code > 299)) {
+      return ::Envoy::Http::FilterHeadersStatus::Continue;
+    }
+    // Apply header mutations.
+    response_header_parser_->evaluateHeaders(headers, encoder_callbacks->streamInfo());
+    // Modify response status code.
+    if (status_code_.has_value()) {
+      auto const code = *status_code_;
+      headers.setStatus(std::to_string(enumToInt(code)));
+      encoder_callbacks->streamInfo().setResponseCode(static_cast<uint32_t>(code));
+    }
+    return ::Envoy::Http::FilterHeadersStatus::Continue;
+  }
+
+  // Modify the request headers & recreate stream.
+  if (custom_response_filter.onLocalReplyCalled()) {
+    // This condition is true if send local reply is called at any point before
+    // the encode call for the custom response filter.
+    // TODO(pradeepcrao): Currently redirect policy is not compatible with send local reply as we
+    // cannot call recreateStream once sendLocalReply has been called.
+    return ::Envoy::Http::FilterHeadersStatus::Continue;
+  }
+  auto downstream_headers = custom_response_filter.downstreamHeaders();
+  RELEASE_ASSERT(downstream_headers != nullptr, "downstream_headers cannot be nullptr");
+
+  ::Envoy::Http::Utility::Url absolute_url;
+  const std::string uri(absl::StrCat(host_, path_));
+  if (!absolute_url.initialize(uri, false)) {
+    IS_ENVOY_BUG(absl::StrCat("Redirect for custom response failed: invalid location {}",
+                              absl::StrCat(host_, path_)));
+    return ::Envoy::Http::FilterHeadersStatus::Continue;
+  }
+  // Don't change the scheme from the original request
+  const bool scheme_is_http = schemeIsHttp(*downstream_headers, decoder_callbacks->connection());
+
+  // Cache original host and path
+  const std::string original_host(downstream_headers->getHostValue());
+  const std::string original_path(downstream_headers->getPathValue());
+  const bool scheme_is_set = (downstream_headers->Scheme() != nullptr);
+  // Replace the original host, scheme and path.
+  Cleanup restore_original_headers(
+      [original_host, original_path, scheme_is_set, scheme_is_http, downstream_headers]() {
+        downstream_headers->setHost(original_host);
+        downstream_headers->setPath(original_path);
+        if (scheme_is_set) {
+          downstream_headers->setScheme(scheme_is_http
+                                            ? ::Envoy::Http::Headers::get().SchemeValues.Http
+                                            : ::Envoy::Http::Headers::get().SchemeValues.Https);
+        }
+      });
+
+  downstream_headers->setScheme(absolute_url.scheme());
+  downstream_headers->setHost(absolute_url.hostAndPort());
+
+  auto path_and_query = path_;
+  // Strip the #fragment from Location URI if it is present.
+  auto fragment_pos = path_.find('#');
+  path_and_query = path_.substr(0, fragment_pos);
```

Comment:
Is this used somewhere?

Reply:
Made the comment clearer. Envoy will reject the request if the `#fragment` is present for the internal redirect.